### ISSUE TITLE
`map_entry` lint fires even when else block takes ownership of key

### ISSUE THREAD

**Issue by TogarashiPepper** [rust-lang/rust-clippy#15233]
### Summary

The `map_entry` lint suggests using the `Entry` API in some cases where it is not possible (without cloning the value, which would be undesirable). In this case the ownership of the value `string` is needed in both the `if` *and* the `else` blocks which makes using the `Entry` API undesirable since the else block would end up being the `Occupied` variant of the match and the `OccupiedEntry` struct doesn't provide a way to get the key we passed to `.entry` back.

### Lint Name

map_entry

### Reproducer

I tried this code (note drop is just a placeholder for any function that needs ownership):

```rust
use std::collections::HashMap;

fn main() {
    let mut map: HashMap<String, u64> = HashMap::new();
    let string = String::new();

    if !map.contains_key(&string) {
        map.insert(string, 42);
    }
    else {
        drop(string);
    }
}
```

I saw this happen:

```
warning: usage of `contains_key` followed by `insert` on a `HashMap`
  --> src/main.rs:7:5
   |
7  | /     if !map.contains_key(&string) {
8  | |         map.insert(string, 42);
9  | |     }
10 | |     else {
11 | |         drop(string);
12 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_entry
   = note: `#[warn(clippy::map_entry)]` on by default
```

I expected to see this happen:
No warning emitted

### Version

```text
rustc 1.88.0 (6b00bc388 2025-06-23)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: aarch64-apple-darwin
release: 1.88.0
LLVM version: 20.1.5
```

**Comment by odysa** [comment]
The current impl throws an warning in this case
https://github.com/rust-lang/rust-clippy/blob/6f2567d16da654e38404c93307594a04a03deddb/clippy_lints/src/entry.rs#L101-L105

Maybe we can fix it by
```rust
if then_search.is_key_used_and_no_copy && else_search.is_key_used_and_no_copy {
    skip();
}
```

**Comment by TogarashiPepper** [comment]
~~Yeah I believe this would resolve this specific issue, but I don't think it generalizes to the case where the `OccupiedEntry` variant of the branch requires ownership. Although I don't know if checking that sort of thing is out of scope for a clippy lint~~
Never mind this case is already handled so actually this fix does completely resolve it

**Comment by odysa** [comment]
The proposed fix didn't work because
`then_search.is_key_used_and_no_copy = false` 
`else_search.is_key_used_and_no_copy = true`

Let me further investigate.

It's related to https://github.com/rust-lang/rust-clippy/issues/9925
